File: internal/sessions/store.go
package sessions

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/gofrs/flock"
)

// Store manages session persistence with in-memory caching and file locking
type Store struct {
	config      *StoreConfig
	cache       *StoreCache
	cacheMu     sync.RWMutex
	lockDir     string
	enableCache bool

	// Cached file mtime with periodic refresh
	mtimeCache    int64
	mtimeCacheMu  sync.RWMutex
	mtimeCacheExp time.Time
}

// StoreCache holds cached session data
type StoreCache struct {
	store     map[string]*SessionEntry
	loadedAt  time.Time
	mtimeMs   int64
	validOnce bool
}

// ReadonlyStore provides a read-only view of the session store
type ReadonlyStore struct {
	store map[string]*SessionEntry
	mu    sync.RWMutex
}

// NewStore creates a new session store
func NewStore(config *StoreConfig) *Store {
	if config == nil {
		panic("config cannot be nil")
	}

	// Ensure store directory exists
	storeDir := filepath.Dir(config.StorePath)
	if err := os.MkdirAll(storeDir, 0755); err != nil {
		log.Printf("[SessionStore] Failed to create store directory: %v", err)
	}

	// Lock directory in the same location as the store
	lockDir := storeDir

	return &Store{
		config:      config,
		lockDir:     lockDir,
		enableCache: config.CacheTTL > 0,
	}
}

// Load loads the session store from disk (with cache support)
// Returns a ReadonlyStore for efficient read-only access
func (s *Store) Load() (map[string]*SessionEntry, error) {
	// Check cache first
	if s.enableCache {
		s.cacheMu.RLock()
		if s.cache != nil && s.isCacheValid(s.cache) {
			// Check if file hasn't been modified (using cached mtime)
			if s.getFileMtimeMsCached() == s.cache.mtimeMs {
				// Return a copy for backward compatibility
				result := s.copyStore(s.cache.store)
				s.cacheMu.RUnlock()
				log.Printf("[SessionStore] Loaded from cache (%d sessions)", len(result))
				return result, nil
			}
		}
		s.cacheMu.RUnlock()
	}

	// Load from disk
	store := make(map[string]*SessionEntry)
	data, err := os.ReadFile(s.config.StorePath)
	if err != nil {
		if os.IsNotExist(err) {
			// First run - return empty store
			log.Printf("[SessionStore] No existing store, starting fresh")
			return store, nil
		}
		return nil, fmt.Errorf("failed to read store: %w", err)
	}

	if err := json.Unmarshal(data, &store); err != nil {
		log.Printf("[SessionStore] Failed to parse store, starting fresh: %v", err)
		return make(map[string]*SessionEntry), nil
	}

	// Update cache
	if s.enableCache {
		s.cacheMu.Lock()
		s.cache = &StoreCache{
			store:    s.copyStore(store),
			loadedAt: time.Now(),
			mtimeMs:  s.getFileMtimeMsCached(),
		}
		s.cacheMu.Unlock()
	}

	log.Printf("[SessionStore] Loaded from disk (%d sessions)", len(store))
	return store, nil
}

// LoadReadonly loads the session store and returns a read-only view
// This is more efficient than Load() for read-heavy workloads
func (s *Store) LoadReadonly() (*ReadonlyStore, error) {
	store, err := s.Load()
	if err != nil {
		return nil, err
	}
	return &ReadonlyStore{store: store}, nil
}

// Get returns a session entry by key from the readonly store
func (r *ReadonlyStore) Get(key string) *SessionEntry {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.store[key]
}

// Has checks if a session exists
func (r *ReadonlyStore) Has(key string) bool {
	r.mu.RLock()
	defer r.mu.RUnlock()
	_, ok := r.store[key]
	return ok
}

// Count returns the number of sessions
func (r *ReadonlyStore) Count() int {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return len(r.store)
}

// All returns a copy of all sessions
func (r *ReadonlyStore) All() map[string]*SessionEntry {
	r.mu.RLock()
	defer r.mu.RUnlock()
	result := make(map[string]*SessionEntry, len(r.store))
	for k, v := range r.store {
		if v != nil {
			copy := *v
			result[k] = &copy
		}
	}
	return result
}

// Save saves the session store to disk (with locking)
func (s *Store) Save(store map[string]*SessionEntry) error {
	return s.withLock(func() error {
		return s.saveUnlocked(store)
	})
}

// Update atomically updates the session store
func (s *Store) Update(mutator func(map[string]*SessionEntry) error) error {
	return s.withLock(func() error {
		// Always re-read inside the lock to avoid clobbering concurrent writers
		store, err := s.loadUnlocked()
		if err != nil {
			return err
		}

		if err := mutator(store); err != nil {
			return err
		}

		return s.saveUnlocked(store)
	})
}

// GetEntry retrieves a single session entry
func (s *Store) GetEntry(sessionKey string) (*SessionEntry, error) {
	store, err := s.Load()
	if err != nil {
		return nil, err
	}
	return store[sessionKey], nil
}

// UpdateEntry updates a single session entry
func (s *Store) UpdateEntry(sessionKey string, update func(*SessionEntry) (*SessionEntry, error)) (*SessionEntry, error) {
	var result *SessionEntry
	err := s.Update(func(store map[string]*SessionEntry) error {
		existing := store[sessionKey]
		patch, err := update(existing)
		if err != nil {
			return err
		}
		if patch == nil {
			result = existing
			return nil
		}

		// Merge patch into existing
		merged := MergeSessionEntry(existing, patch)
		store[sessionKey] = merged
		result = merged
		return nil
	})
	return result, err
}

// RecordInboundMeta records session metadata from an incoming webhook message
func (s *Store) RecordInboundMeta(sessionKey string, webhookMsgID string, deliveryCtx *DeliveryContext) (*SessionEntry, error) {
	return s.UpdateEntry(sessionKey, func(existing *SessionEntry) (*SessionEntry, error) {
		now := time.Now().UnixMilli()

		if existing != nil {
			// Update existing entry
			return &SessionEntry{
				SessionID:        existing.SessionID,
				UpdatedAt:        now,
				SessionFile:      existing.SessionFile,
				DeliveryContext:  deliveryCtx,
				LastChannel:      deliveryChannel(deliveryCtx),
				LastTo:           deliveryTo(deliveryCtx),
				LastAccountId:    deliveryAccountId(deliveryCtx),
				LastThreadId:     deliveryThreadId(deliveryCtx),
				WebhookMessageID: webhookMsgID,
				WebhookSessionID: sessionKey,
				// Preserve existing state
				SystemSent:       existing.SystemSent,
				AbortedLastRun:   existing.AbortedLastRun,
				ThinkingLevel:    existing.ThinkingLevel,
				VerboseLevel:     existing.VerboseLevel,
				ReasoningLevel:   existing.ReasoningLevel,
				SendPolicy:       existing.SendPolicy,
				ModelOverride:    existing.ModelOverride,
				ProviderOverride: existing.ProviderOverride,
			}, nil
		}

		// Create new entry - minimal state for webhook sessions
		return &SessionEntry{
			SessionID:        generateSessionID(),
			UpdatedAt:        now,
			DeliveryContext:  deliveryCtx,
			LastChannel:      deliveryChannel(deliveryCtx),
			LastTo:           deliveryTo(deliveryCtx),
			LastAccountId:    deliveryAccountId(deliveryCtx),
			LastThreadId:     deliveryThreadId(deliveryCtx),
			WebhookMessageID: webhookMsgID,
			WebhookSessionID: sessionKey,
		}, nil
	})
}

// UpdateLastRoute updates the last delivery route for a session
func (s *Store) UpdateLastRoute(sessionKey string, deliveryCtx *DeliveryContext) (*SessionEntry, error) {
	return s.UpdateEntry(sessionKey, func(existing *SessionEntry) (*SessionEntry, error) {
		patch := &SessionEntry{
			UpdatedAt:       time.Now().UnixMilli(),
			DeliveryContext: deliveryCtx,
			LastChannel:     deliveryChannel(deliveryCtx),
			LastTo:          deliveryTo(deliveryCtx),
			LastAccountId:   deliveryAccountId(deliveryCtx),
			LastThreadId:    deliveryThreadId(deliveryCtx),
		}
		return patch, nil
	})
}

// loadUnlocked loads without cache checks (must be called with lock held)
func (s *Store) loadUnlocked() (map[string]*SessionEntry, error) {
	store := make(map[string]*SessionEntry)
	data, err := os.ReadFile(s.config.StorePath)
	if err != nil {
		if os.IsNotExist(err) {
			return store, nil
		}
		return nil, fmt.Errorf("failed to read store: %w", err)
	}

	if err := json.Unmarshal(data, &store); err != nil {
		log.Printf("[SessionStore] Failed to parse store: %v", err)
		return make(map[string]*SessionEntry), nil
	}

	return store, nil
}

// saveUnlocked saves without locking (must be called with lock held)
func (s *Store) saveUnlocked(store map[string]*SessionEntry) error {
	// Invalidate cache and mtime cache on write
	s.cacheMu.Lock()
	s.cache = nil
	s.cacheMu.Unlock()

	s.mtimeCacheMu.Lock()
	s.mtimeCache = 0
	s.mtimeCacheExp = time.Time{}
	s.mtimeCacheMu.Unlock()

	// Serialize
	data, err := json.MarshalIndent(store, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal store: %w", err)
	}

	// Write atomically using temp file
	tmpPath := s.config.StorePath + ".tmp"
	if err := os.WriteFile(tmpPath, data, 0600); err != nil {
		return fmt.Errorf("failed to write temp file: %w", err)
	}

	// Rename atomically
	if err := os.Rename(tmpPath, s.config.StorePath); err != nil {
		os.Remove(tmpPath) // Clean up temp file
		return fmt.Errorf("failed to rename store file: %w", err)
	}

	log.Printf("[SessionStore] Saved %d sessions", len(store))
	return nil
}

// isCacheValid checks if a cache entry is still valid
func (s *Store) isCacheValid(cache *StoreCache) bool {
	if cache == nil {
		return false
	}
	return time.Since(cache.loadedAt) < s.config.CacheTTL
}

// copyStore creates a shallow copy of the session store map
// Individual SessionEntry values are copied by value (not deep cloned)
// This is safe because SessionEntry contains only primitive types and pointers
// that are never mutated after being stored
func (s *Store) copyStore(store map[string]*SessionEntry) map[string]*SessionEntry {
	result := make(map[string]*SessionEntry, len(store))
	for k, v := range store {
		if v != nil {
			// Shallow copy - copy the struct but not nested pointers
			// Since DeliveryContext is the only nested pointer and we don't mutate it,
			// this is safe for read-only access
			copy := *v
			result[k] = &copy
		}
	}
	return result
}

// getFileMtimeMsCached gets the file modification time with caching
// Cache expires after 1 second to reduce syscalls while staying fresh
func (s *Store) getFileMtimeMsCached() int64 {
	s.mtimeCacheMu.RLock()
	if time.Now().Before(s.mtimeCacheExp) && s.mtimeCache > 0 {
		mtime := s.mtimeCache
		s.mtimeCacheMu.RUnlock()
		return mtime
	}
	s.mtimeCacheMu.RUnlock()

	// Cache miss or expired, get fresh value
	mtime := s.getFileMtimeMs()

	s.mtimeCacheMu.Lock()
	s.mtimeCache = mtime
	s.mtimeCacheExp = time.Now().Add(time.Second)
	s.mtimeCacheMu.Unlock()

	return mtime
}

// getFileMtimeMs gets the file modification time in milliseconds
func (s *Store) getFileMtimeMs() int64 {
	info, err := os.Stat(s.config.StorePath)
	if err != nil {
		return 0
	}
	return info.ModTime().UnixMilli()
}

// withLock executes a function with the store lock held using flock
func (s *Store) withLock(fn func() error) error {
	lockPath := s.config.StorePath + ".lock"
	timeout := s.config.LockTimeout

	// Ensure lock directory exists
	if err := os.MkdirAll(s.lockDir, 0755); err != nil {
		return fmt.Errorf("failed to create lock directory: %w", err)
	}

	// Use flock for proper file locking
	fileLock := flock.New(lockPath)

	// Try to get lock with timeout using exponential backoff
	startedAt := time.Now()
	pollInterval := 25 * time.Millisecond
	staleDuration := 30 * time.Second

	for {
		locked, err := fileLock.TryLock()
		if err != nil {
			return fmt.Errorf("lock error: %w", err)
		}
		if locked {
			defer fileLock.Unlock()
			// Check for stale lock info and log
			if info, err := os.Stat(lockPath); err == nil {
				age := time.Since(info.ModTime())
				if age > staleDuration {
					log.Printf("[SessionStore] Warning: lock file is %v old (may indicate crashed process)", age)
				}
			}
			return fn()
		}

		// Check timeout
		if time.Since(startedAt) > timeout {
			return fmt.Errorf("timeout acquiring lock: %s", lockPath)
		}

		// Wait before retrying
		time.Sleep(pollInterval)
	}
}

// Helper functions for delivery context
func deliveryChannel(ctx *DeliveryContext) string {
	if ctx != nil {
		return ctx.Channel
	}
	return ""
}

func deliveryTo(ctx *DeliveryContext) string {
	if ctx != nil {
		return ctx.To
	}
	return ""
}

func deliveryAccountId(ctx *DeliveryContext) string {
	if ctx != nil {
		return ctx.AccountId
	}
	return ""
}

func deliveryThreadId(ctx *DeliveryContext) string {
	if ctx != nil {
		return ctx.ThreadId
	}
	return ""
}

// generateSessionID generates a unique session ID
// Using timestamp-based ID for simplicity (can be replaced with UUID)
func generateSessionID() string {
	return fmt.Sprintf("sess_%d", time.Now().UnixNano())
}

// GenerateSessionID generates a unique session ID (exported version)
func GenerateSessionID() string {
	return generateSessionID()
}

File: src/sessions/store.rs
use anyhow::Result;
use fs2::FileExt;
use std::collections::HashMap;
use std::fs::OpenOptions;
use std::io::{Read, Write};
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use super::types::{
    current_timestamp, generate_session_id, DeliveryContext, SessionEntry, SessionStore,
};

/// Store configuration
#[derive(Debug, Clone)]
pub struct StoreConfig {
    pub store_path: PathBuf,
    pub cache_ttl: Duration,
}

impl StoreConfig {
    pub fn new(store_path: PathBuf) -> Self {
        Self {
            store_path,
            cache_ttl: Duration::from_secs(45),
        }
    }
}

/// Session store with file-based persistence
pub struct Store {
    config: StoreConfig,
    cache: Arc<Mutex<CacheEntry>>,
}

struct CacheEntry {
    data: SessionStore,
    loaded_at: Instant,
}

impl Store {
    pub fn new(config: StoreConfig) -> Self {
        Self {
            config,
            cache: Arc::new(Mutex::new(CacheEntry {
                data: HashMap::new(),
                loaded_at: Instant::now() - Duration::from_secs(100), // Force initial load
            })),
        }
    }

    /// Load the session store from disk
    pub fn load(&self) -> Result<SessionStore> {
        // Check cache first
        {
            let cache = self.cache.lock().unwrap();
            if cache.loaded_at.elapsed() < self.config.cache_ttl {
                return Ok(cache.data.clone());
            }
        }

        // Load from disk with file locking
        let file = OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .truncate(true)
            .open(&self.config.store_path)?;

        file.lock_shared()?;

        let mut contents = String::new();
        let mut file_ref = &file;
        file_ref.read_to_string(&mut contents)?;

        let store: SessionStore = if contents.is_empty() {
            HashMap::new()
        } else {
            serde_json::from_str(&contents).unwrap_or_else(|_| HashMap::new())
        };

        file.unlock()?;

        // Update cache
        {
            let mut cache = self.cache.lock().unwrap();
            cache.data = store.clone();
            cache.loaded_at = Instant::now();
        }

        Ok(store)
    }

    /// Save the session store to disk
    fn save(&self, store: &SessionStore) -> Result<()> {
        let file = OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .truncate(true)
            .open(&self.config.store_path)?;

        file.lock_exclusive()?;

        let contents = serde_json::to_string_pretty(store)?;
        let mut file_ref = &file;
        file_ref.write_all(contents.as_bytes())?;

        file.unlock()?;

        // Update cache
        {
            let mut cache = self.cache.lock().unwrap();
            cache.data = store.clone();
            cache.loaded_at = Instant::now();
        }

        Ok(())
    }

    /// Update store using a callback function
    pub fn update<F>(&self, f: F) -> Result<()>
    where
        F: FnOnce(&mut SessionStore) -> Result<()>,
    {
        let mut store = self.load()?;
        f(&mut store)?;
        self.save(&store)?;
        Ok(())
    }

    /// Get a session entry
    #[allow(dead_code)]
    pub fn get_entry(&self, key: &str) -> Result<Option<SessionEntry>> {
        let store = self.load()?;
        Ok(store.get(key).cloned())
    }

    /// Delete a session entry. Returns true if a row was removed.
    #[allow(dead_code)]
    pub fn delete_entry(&self, key: &str) -> Result<bool> {
        let mut removed = false;
        self.update(|store| {
            removed = store.remove(key).is_some();
            Ok(())
        })?;
        Ok(removed)
    }

    /// Find the (key, entry) pair whose session_id matches.
    /// Scans the full store; O(n). Used for `session.get` by id lookup.
    #[allow(dead_code)]
    pub fn find_by_session_id(&self, session_id: &str) -> Result<Option<(String, SessionEntry)>> {
        let store = self.load()?;
        Ok(store
            .iter()
            .find(|(_, v)| v.session_id == session_id)
            .map(|(k, v)| (k.clone(), v.clone())))
    }

    /// Update a session entry using a callback
    pub fn update_entry<F>(&self, key: &str, f: F) -> Result<SessionEntry>
    where
        F: FnOnce(Option<&SessionEntry>) -> Result<SessionEntry>,
    {
        let mut result = None;
        self.update(|store| {
            let existing = store.get(key);
            let entry = f(existing)?;
            store.insert(key.to_string(), entry.clone());
            result = Some(entry);
            Ok(())
        })?;
        Ok(result.unwrap())
    }

    /// Record inbound message metadata
    pub fn record_inbound_meta(
        &self,
        session_key: &str,
        message_id: &str,
        delivery_ctx: &DeliveryContext,
    ) -> Result<SessionEntry> {
        self.update_entry(session_key, |existing| {
            let mut entry = if let Some(e) = existing {
                e.clone()
            } else {
                SessionEntry {
                    session_id: generate_session_id(),
                    updated_at: current_timestamp(),
                    session_file: None,
                    delivery_context: None,
                    last_channel: None,
                    last_to: None,
                    last_account_id: None,
                    last_thread_id: None,
                    webhook_message_id: None,
                    webhook_session_id: None,
                }
            };

            // Update metadata
            entry.updated_at = current_timestamp();
            entry.delivery_context = Some(delivery_ctx.clone());
            entry.last_channel = delivery_ctx.channel.clone();
            entry.last_to = delivery_ctx.to.clone();
            entry.last_account_id = delivery_ctx.account_id.clone();
            entry.last_thread_id = delivery_ctx.thread_id.clone();
            entry.webhook_message_id = Some(message_id.to_string());

            Ok(entry)
        })
    }

    /// Update last route for a session
    #[allow(dead_code)]
    pub fn update_last_route(
        &self,
        session_key: &str,
        delivery_ctx: &DeliveryContext,
    ) -> Result<SessionEntry> {
        self.update_entry(session_key, |existing| {
            let mut entry = if let Some(e) = existing {
                e.clone()
            } else {
                SessionEntry {
                    session_id: generate_session_id(),
                    updated_at: current_timestamp(),
                    session_file: None,
                    delivery_context: None,
                    last_channel: None,
                    last_to: None,
                    last_account_id: None,
                    last_thread_id: None,
                    webhook_message_id: None,
                    webhook_session_id: None,
                }
            };

            // Update last route
            entry.updated_at = current_timestamp();
            if delivery_ctx.channel.is_some() {
                entry.last_channel = delivery_ctx.channel.clone();
            }
            if delivery_ctx.account_id.is_some() {
                entry.last_account_id = delivery_ctx.account_id.clone();
            }

            Ok(entry)
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::env;
    use std::path::PathBuf;

    fn temp_store(name: &str) -> Store {
        let mut path: PathBuf = env::temp_dir();
        path.push(format!(
            "openclaw-sessions-test-{}-{}.json",
            name,
            uuid::Uuid::new_v4()
        ));
        let _ = std::fs::remove_file(&path);
        Store::new(StoreConfig::new(path))
    }

    fn ctx(channel: &str, to: &str) -> DeliveryContext {
        DeliveryContext {
            channel: Some(channel.to_string()),
            to: Some(to.to_string()),
            account_id: Some("uid-test".to_string()),
            thread_id: None,
        }
    }

    #[test]
    fn record_and_get_entry_round_trips() {
        let store = temp_store("rt");
        let entry = store
            .record_inbound_meta("webhook:msg-1", "msg-1", &ctx("webhook", "msg-1"))
            .expect("record");
        let got = store.get_entry("webhook:msg-1").expect("get");
        assert_eq!(got.unwrap().session_id, entry.session_id);
    }

    #[test]
    fn delete_entry_returns_true_when_present() {
        let store = temp_store("del-yes");
        store
            .record_inbound_meta("k1", "m1", &ctx("webhook", "m1"))
            .unwrap();
        assert!(store.delete_entry("k1").unwrap());
        assert!(store.get_entry("k1").unwrap().is_none());
    }

    #[test]
    fn delete_entry_returns_false_when_absent() {
        let store = temp_store("del-no");
        assert!(!store.delete_entry("nope").unwrap());
    }

    #[test]
    fn find_by_session_id_returns_owner_key() {
        let store = temp_store("find");
        store
            .record_inbound_meta("webhook:a", "a", &ctx("webhook", "a"))
            .unwrap();
        let entry = store
            .record_inbound_meta("webhook:b", "b", &ctx("webhook", "b"))
            .unwrap();
        let (key, found) = store
            .find_by_session_id(&entry.session_id)
            .unwrap()
            .unwrap();
        assert_eq!(key, "webhook:b");
        assert_eq!(found.session_id, entry.session_id);
    }

    #[test]
    fn find_by_session_id_returns_none_when_missing() {
        let store = temp_store("find-miss");
        assert!(store.find_by_session_id("nope").unwrap().is_none());
    }
}
